segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/StorageWriterFactory.java

```diff
@@ -86,6 +92,27 @@ public int getId() {
             return this.operationLog.truncate(upToSequence, timeout);
         }
 
+        @Override
+        public CompletableFuture<Void> persistAttributes(long streamSegmentId, Map<UUID, Long> attributes, Duration timeout) {
+            TimeoutTimer timer = new TimeoutTimer(timeout);
+            return this.attributeIndex
+                    .forSegment(streamSegmentId, timer.getRemaining())
+                    .thenCompose(ai -> ai.put(attributes, timer.getRemaining()));
```

Comment:
Is this the primary call that we make to `SegmentAttributeIndex.put`  to populate the attribute index of a segment?

Reply:
Yes. The `StorageWriter`'s `SegmentAggregator`s will accumulate a number of attribute updates in their own buffers (just like they do with regular data), and will invoke this method to persist them into the SegmentAttributeIndex/Tier2. Only after both this and the regular data write to tier2 will the `SegmentAggregator`s acknowledge that those operations have been durably persisted in Tier2.